MINIFICPP-1092 - Make CoAP compile and work on Windows

bakaid opened the pull request:
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

arpadboda commented on extensions/coap/server/CoapServer.h:
```diff
@@ -18,24 +18,26 @@
-  POST,
-  PUT,
-  DELETE
+enum class Method {
```
👍

arpadboda commented on cmake/BundledLibCoAP.cmake:
```diff
@@ -0,0 +1,84 @@
+        ExternalProject_Add(
+                coap-external
+                URL "https://github.com/obgm/libcoap/archive/v4.2.1.tar.gz"
+                URL_HASH "SHA256=29a0394a265d3febee41e5e2dc03d34292a0aede37f5f80334e529ac0dab2321"
```
Can we move this and the URL hash to a variable to avoid updating only for one platform?

bakaid replied:
Fair point, will fix.

phrocker commented:
This is awesome. I'll try to split some time to check but I had some compile time errors as this could be my instance. I'll spin up another windows instance on windows 10 home server as my current one may have an unrelated issue...but I see appveyor errors. are those related?

bakaid commented:
> This is awesome. I'll try to split some time to check but I had some compile time errors as this could be my instance. I'll spin up another windows instance on windows 10 home server as my current one may have an unrelated issue...but I see appveyor errors. are those related?

I have compiled and tried it on a 64-bit Windows 10 machine. The AppVeyor error seems unrealted: `CUSTOMBUILD : error : timer thread did not terminate timely [C:\projects\nifi-minifi-cpp\build\curl-external.vcxproj]`

phrocker commented:
> > This is awesome. I'll try to split some time to check but I had some compile time errors as this could be my instance. I'll spin up another windows instance on windows 10 home server as my current one may have an unrelated issue...but I see appveyor errors. are those related?
> 
> I have compiled and tried it on a 64-bit Windows 10 machine. The AppVeyor error seems unrelated: `CUSTOMBUILD : error : timer thread did not terminate timely [C:\projects\nifi-minifi-cpp\build\curl-external.vcxproj]`

Cool. I'll run the build across variants of 10, 2012 server, 2016 server, and 2019 server later today as I usually do.

phrocker commented:
We may/should add those variants to appveyor if they have licensing to support that. I have paid for those licenses in the past -- but I imagine apache has access through appveyor.

phrocker commented:
@bakaid  I saw some updates in my email and have been jumping around between projects, so I don't want to be the long pole. I have it on my white board to  re-try on windows * server, so if you don't get a chance to try it I'd go ahead and merge. I'm good with the changes. We can always make a follow on task if issues arise.

bakaid commented:
@arpadboda I've rebased this to the new master and added an ObjectFactoryInitializer for the CoAP extension to properly initialize winsock.

szaszm commented on extensions/coap/COAPLoader.cpp:
```diff
@@ -18,7 +18,28 @@
+bool COAPObjectFactoryInitializer::initialize() {
+#ifdef WIN32
+  static WSADATA s_wsaData;
+  int iWinSockInitResult = WSAStartup(MAKEWORD(2, 2), &s_wsaData);
```
Did you consider logging the issue in case of failed `WSAStartup`?

bakaid replied:
`WSAStartup` is extremely unlikely to fail, that's why I didn't add one originally, but it won't hurt, so I've added it now.

szaszm commented on extensions/coap/tests/CMakeLists.txt:
```diff
@@ -40,7 +40,7 @@ FOREACH(testfile ${CURL_INTEGRATION_TESTS})
 	target_include_directories(${testfilename} BEFORE PRIVATE "${CMAKE_SOURCE_DIR}/extensions/civetweb/")
 	target_include_directories(${testfilename} BEFORE PRIVATE ./include)
-    createTests("${testfilename}")
+        createTests("${testfilename}")
```
The indentation is still inconsistent: the rest of the file uses tabs while this line now uses 8 spaces instead of 4.

bakaid replied:
Yep, that's fair, edited outside my normal IDE, fixed.

bakaid commented:
@arpadboda rebased to fresh master, addressed all review comments, built and ran tests on macOS and Windows. Let's wait for CI, but otherwise I think this should be good to go.

szaszm commented:
my +1 still applies, fyi

arpadboda commented on extensions/coap/COAPLoader.cpp:
```diff
@@ -16,9 +16,36 @@
+  if (iWinSockInitResult != 0) {
+    logging::LoggerFactory<COAPObjectFactoryInitializer>::getLogger()->log_error("WSAStartup failed with error %d", iWinSockInitResult);
+    return false;
+  } else {
```
Be  removing the else (which is fine because of the return above) we can also remove the else part of the ifdef, but it's just fancy, this should work fine.

bakaid replied:
I think this is better readable this way.

arpadboda commented on extensions/coap/COAPLoader.h:
```diff
@@ -28,6 +28,13 @@
 
+class COAPObjectFactoryInitializer : public core::ObjectFactoryInitializer {
+ public:
+  virtual bool initialize();
```
Could you add override here?
I guess this PR was created before we agreed on using override.

bakaid replied:
Done.
